src/core/mac/data_poll_handler.cpp

```diff
@@ -273,10 +273,10 @@ void DataPollHandler::HandleSentFrame(const Mac::TxFrame &aFrame, Error aError,
                 uint32_t frameCounter;
                 uint8_t  keyId;
 
-                IgnoreError(aFrame.GetFrameCounter(frameCounter));
+                OT_ASSERT(aFrame.GetFrameCounter(frameCounter) == kErrorNone);
```

Comment:
Use of `OT_ASSERT()` like this is not correct and won't work properly on all builds.
Note that `OT_ASSERT()` may be configured to compile out as NOOP (empty). Then we won't get the frame counter value.

Reply:
ok, I will update it to VerifyOrExit .
I realize that there is not much we can do in such case and originally I was going to mark it as "Intended", but I realized that leaving it as is may lead to sending some not-so-random data from the stack that may happen to be sensitive.

Reply:
I think `OT_ASSERT` is appropriate because we want to ensure that transmitted frames are always formed properly. However, to @abtink 's point, we have to separate the logic from verification into separate lines:
```
error = aFrame.GetFrameCounter(frameCounter);
OT_ASSERT(error == kErrorNone);
```

Reply:
I agree with @jwhui and I think assert would be better here (helps detect any issue with platform implementation since we expect the tx done callback to give back the same frame as the one given by OT stack which should be well-formed). 

We just need to seperate the logic. Another thing to note, since `OT_ASSERT()` can become NOOP, and we are adding a new local `Error error` variable that is only used in the assert check, we need `OT_UNUSED_VARIABLE(error);` to silence the warning that `error` is unused in the configs where assert is disabled. You can check `RandomManager::RandomManager(void)` or `CoapSecure::SetPsk()` which use very similar pattern.

Reply:
Sorry. My bad, I did not realize that assert evaluates to nothing and didn't get you the first time. Will fix that on Monday.